File: client/lib/api/v2/schemas/rubrics.ts
/**
 * Rubrics V2 API Schemas
 * Schema definitions for rubrics v2 endpoints with hierarchical structure
 */

import { z } from "zod";
import { DepartmentMappingSchema } from "./base";

// Custom mapping schemas for rubrics (include points unlike base)
export const StandardGroupMappingItemSchema = z.object({
  name: z.string(),
  description: z.string(),
  points: z.number(),
  passPoints: z.number(),
});

export const StandardMappingItemSchema = z.object({
  name: z.string(),
  description: z.string(),
  points: z.number(),
});

export const StandardGroupMappingDetailSchema = z.object({
  name: z.string(),
  description: z.string(),
});

export const StandardGroupsMappingSchema = z.record(
  z.string(),
  StandardGroupMappingItemSchema
);
export const StandardsMappingSchema = z.record(
  z.string(),
  StandardMappingItemSchema
);
export const StandardGroupsMappingDetailSchema = z.record(
  z.string(),
  StandardGroupMappingDetailSchema
);

export type StandardGroupsMapping = z.infer<typeof StandardGroupsMappingSchema>;
export type StandardsMapping = z.infer<typeof StandardsMappingSchema>;

// ============================================================================
// REQUEST SCHEMAS
// ============================================================================

export const RubricsFiltersSchema = z.object({
  departmentIds: z.array(z.string()),
  profileId: z.string(),
});

export type RubricsFilters = z.infer<typeof RubricsFiltersSchema>;

// ============================================================================
// RESPONSE SCHEMAS
// ============================================================================

// Rubric item (normalized - IDs only with hierarchical structure)
export const RubricItemSchema = z.object({
  rubric_id: z.string(),
  name: z.string(),
  description: z.string(),
  points: z.number(),
  passPoints: z.number(),
  default_rubric: z.boolean(),
  can_edit: z.boolean(),
  can_delete: z.boolean(),
  can_duplicate: z.boolean(),
  // Hierarchical structure: standard_group_id -> array of standard_ids
  standard_groups: z.record(z.string(), z.array(z.string())),
});

export const RubricsListResponseSchema = z.object({
  rubrics: z.array(RubricItemSchema),
  standard_groups_mapping: StandardGroupsMappingSchema,
  standards_mapping: StandardsMappingSchema,
});

export type RubricsListResponse = z.infer<typeof RubricsListResponseSchema>;
export type RubricItem = z.infer<typeof RubricItemSchema>;

// ============================================================================
// DETAIL SCHEMAS
// ============================================================================

export const RubricDetailRequestSchema = z.object({
  rubricId: z.string(),
  profileId: z.string(),
});

export type RubricDetailRequest = z.infer<typeof RubricDetailRequestSchema>;

export const StandardGroupDetailSchema = z.object({
  points: z.number(),
  passPoints: z.number(),
  standard_ids: z.array(z.string()),
});

export type StandardGroupDetail = z.infer<typeof StandardGroupDetailSchema>;

export const RubricDetailResponseSchema = z.object({
  // Basic fields
  name: z.string(),
  description: z.string(),
  department_id: z.string(),
  valid_department_ids: z.array(z.string()),
  points: z.number(),
  passPoints: z.number(),
  active: z.boolean(),
  default_rubric: z.boolean(),
  can_edit: z.boolean(), // Permission flag for editing

  // Standard groups structure
  standard_group_ids: z.array(z.string()),
  standard_groups_detail: z.record(z.string(), StandardGroupDetailSchema),

  // Top-level mappings
  standard_groups_mapping: StandardGroupsMappingDetailSchema,
  standards_mapping: StandardsMappingSchema,
  department_mapping: DepartmentMappingSchema,
});

export type RubricDetailResponse = z.infer<typeof RubricDetailResponseSchema>;

// Default detail request
export const RubricDetailDefaultRequestSchema = z.object({
  profileId: z.string(),
});

export type RubricDetailDefaultRequest = z.infer<
  typeof RubricDetailDefaultRequestSchema
>;

// ============================================================================
// MUTATION SCHEMAS
// ============================================================================

// Nested standard schema for create/update
export const StandardCreateSchema = z.object({
  name: z.string(),
  description: z.string(),
  points: z.number(),
});

export type StandardCreate = z.infer<typeof StandardCreateSchema>;

// Standard update schema for incremental updates
export const StandardUpdateSchema = z.object({
  id: z.string().optional(), // Present for updates, absent for creates
  name: z.string(),
  description: z.string(),
  points: z.number(),
  deleted: z.boolean().default(false), // Mark for deletion
});

export type StandardUpdate = z.infer<typeof StandardUpdateSchema>;

// Nested standard group schema for create/update
export const StandardGroupCreateSchema = z.object({
  name: z.string(),
  short_name: z.string(),
  description: z.string(),
  points: z.number(),
  passPoints: z.number(),
  standards: z.array(StandardCreateSchema),
});

export type StandardGroupCreate = z.infer<typeof StandardGroupCreateSchema>;

// Standard group update schema for incremental updates
export const StandardGroupUpdateSchema = z.object({
  id: z.string().optional(), // Present for updates, absent for creates
  name: z.string(),
  short_name: z.string(),
  description: z.string(),
  points: z.number(),
  passPoints: z.number(),
  standards: z.array(StandardUpdateSchema),
  deleted: z.boolean().default(false), // Mark for deletion
});

export type StandardGroupUpdate = z.infer<typeof StandardGroupUpdateSchema>;

// Create request
export const CreateRubricRequestSchema = z.object({
  name: z.string(),
  description: z.string(),
  department_id: z.string(),
  active: z.boolean(),
  default_rubric: z.boolean(),
  points: z.number(),
  passPoints: z.number(),
  standard_groups: z.array(StandardGroupCreateSchema),
});

export type CreateRubricRequest = z.infer<typeof CreateRubricRequestSchema>;

export const CreateRubricResponseSchema = z.object({
  success: z.boolean(),
  rubricId: z.string(),
  message: z.string(),
});

export type CreateRubricResponse = z.infer<typeof CreateRubricResponseSchema>;

// Update request - uses StandardGroupUpdate for incremental updates
export const UpdateRubricRequestSchema = z.object({
  rubricId: z.string(),
  name: z.string(),
  description: z.string(),
  department_id: z.string(),
  active: z.boolean(),
  default_rubric: z.boolean(),
  standard_groups: z.array(StandardGroupUpdateSchema),
  // points and passPoints are auto-calculated by server
});

export type UpdateRubricRequest = z.infer<typeof UpdateRubricRequestSchema>;

export const UpdateRubricResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
  points: z.number(), // Auto-calculated from standard groups
  passPoints: z.number(), // Auto-calculated from standard groups
});

export type UpdateRubricResponse = z.infer<typeof UpdateRubricResponseSchema>;

// Duplicate request
export const DuplicateRubricRequestSchema = z.object({
  rubricId: z.string(),
});

export type DuplicateRubricRequest = z.infer<
  typeof DuplicateRubricRequestSchema
>;

export const DuplicateRubricResponseSchema = z.object({
  success: z.boolean(),
  rubricId: z.string(),
  message: z.string(),
});

export type DuplicateRubricResponse = z.infer<
  typeof DuplicateRubricResponseSchema
>;

// Delete request
export const DeleteRubricRequestSchema = z.object({
  rubricId: z.string(),
});

export type DeleteRubricRequest = z.infer<typeof DeleteRubricRequestSchema>;

export const DeleteRubricResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type DeleteRubricResponse = z.infer<typeof DeleteRubricResponseSchema>;

File: server/app/schemas/rubrics.py
"""Rubrics V2 API schemas with hierarchical structure."""

from typing import Dict, List

from pydantic import BaseModel

from .base import DepartmentMapping, StandardGroupsMapping, StandardsMapping

# ============================================================================
# CENTRALIZED MAPPING TYPES
# ============================================================================


class StandardGroupMappingItem(BaseModel):
    """Standard group mapping item for list response."""

    name: str
    description: str
    points: int
    passPoints: int


class StandardMappingItem(BaseModel):
    """Standard mapping item."""

    name: str
    description: str
    points: int


class StandardGroupMappingDetail(BaseModel):
    """Standard group mapping for detail response."""

    name: str
    description: str


# ============================================================================
# REQUEST SCHEMAS
# ============================================================================


class RubricsFilters(BaseModel):
    """Filters for rubrics list."""

    departmentIds: List[str]
    profileId: str


# ============================================================================
# RESPONSE SCHEMAS
# ============================================================================


class RubricItem(BaseModel):
    """Rubric item in list response with hierarchical structure."""

    rubric_id: str
    name: str
    description: str
    points: int
    passPoints: int
    can_edit: bool
    can_delete: bool
    can_duplicate: bool
    # Hierarchical: standard_group_id -> array of standard_ids
    standard_groups: Dict[str, List[str]]


class RubricsListResponse(BaseModel):
    """Response for rubrics list endpoint."""

    rubrics: List[RubricItem]
    standard_groups_mapping: StandardGroupsMapping
    standards_mapping: StandardsMapping


# ============================================================================
# DETAIL SCHEMAS
# ============================================================================


class RubricDetailRequest(BaseModel):
    """Request for rubric detail."""

    rubricId: str
    profileId: str


class StandardGroupDetail(BaseModel):
    """Standard group detail for detail response."""

    points: int
    passPoints: int
    standard_ids: List[str]


class RubricDetailResponse(BaseModel):
    """Response for rubric detail endpoint."""

    # Basic fields
    name: str
    description: str
    department_id: str
    valid_department_ids: List[str]
    points: int
    passPoints: int
    active: bool
    default_rubric: bool

    # Standard groups structure
    standard_group_ids: List[str]
    standard_groups_detail: Dict[str, StandardGroupDetail]

    # Top-level mappings
    standard_groups_mapping: StandardGroupsMapping
    standards_mapping: StandardsMapping
    department_mapping: DepartmentMapping


class RubricDetailDefaultRequest(BaseModel):
    """Request for default rubric detail."""

    profileId: str


# ============================================================================
# MUTATION SCHEMAS - NESTED STRUCTURES
# ============================================================================


class StandardCreate(BaseModel):
    """Standard creation schema."""

    name: str
    description: str
    points: int


class StandardGroupCreate(BaseModel):
    """Standard group creation schema with nested standards."""

    name: str
    short_name: str
    description: str
    points: int
    passPoints: int
    standards: List[StandardCreate]


class CreateRubricRequest(BaseModel):
    """Request to create rubric with nested structure."""

    name: str
    description: str
    department_id: str
    active: bool
    default_rubric: bool
    points: int
    passPoints: int
    standard_groups: List[StandardGroupCreate]


class CreateRubricResponse(BaseModel):
    """Response from create rubric."""

    success: bool
    rubricId: str
    message: str


class UpdateRubricRequest(BaseModel):
    """Request to update rubric with nested structure."""

    rubricId: str
    name: str
    description: str
    department_id: str
    active: bool
    default_rubric: bool
    points: int
    passPoints: int
    standard_groups: List[StandardGroupCreate]


class UpdateRubricResponse(BaseModel):
    """Response from update rubric."""

    success: bool
    message: str


class DuplicateRubricRequest(BaseModel):
    """Request to duplicate rubric."""

    rubricId: str


class DuplicateRubricResponse(BaseModel):
    """Response from duplicate rubric."""

    success: bool
    rubricId: str
    message: str


class DeleteRubricRequest(BaseModel):
    """Request to delete rubric."""

    rubricId: str


class DeleteRubricResponse(BaseModel):
    """Response from delete rubric."""

    success: bool
    message: str

